// @redocly/openapi-cli/lib/__tests__/commands/push.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const push_1 = require("../../commands/push");
jest.mock('fs');
jest.mock('node-fetch');
jest.mock('@redocly/openapi-core');
jest.mock('../../utils');
describe('push', () => {
    const redoclyClient = require('@redocly/openapi-core').__redoclyClient;
    beforeAll(() => {
        jest.spyOn(process.stdout, 'write').mockImplementation(() => true);
    });
    it('pushes definition', () => __awaiter(void 0, void 0, void 0, function* () {
        yield push_1.handlePush({
            upsert: true,
            entrypoint: 'spec.json',
            destination: '@org/my-api@1.0.0',
            branchName: 'test',
        });
        expect(redoclyClient.registryApi.prepareFileUpload).toBeCalledTimes(1);
        expect(redoclyClient.registryApi.pushApi).toBeCalledTimes(1);
        expect(redoclyClient.registryApi.pushApi).toHaveBeenLastCalledWith({
            branch: 'test',
            filePaths: ['filePath'],
            isUpsert: true,
            name: 'my-api',
            organizationId: 'org',
            rootFilePath: 'filePath',
            version: '1.0.0',
        });
    }));
});
